src/shared/machine-id-setup.c

```diff
@@ -48,13 +48,22 @@ static int acquire_machine_id_from_credential(sd_id128_t *ret) {
         return 0;
 }
 
-static int generate_machine_id(const char *root, sd_id128_t *ret) {
+static int acquire_machine_id(const char *root, sd_id128_t *ret) {
         _cleanup_close_ int fd = -EBADF;
         int r;
 
         assert(ret);
 
-        /* First, try reading the D-Bus machine id, unless it is a symlink */
+        /* First, try reading the machine ID from /run/machine-id, which may not be mounted on
+         * /etc/machine-id yet. This is important on switching root, Otherwise, machine ID may be changed
+         * after the transition. */
```

Comment:
hmm, when we run in the initrd, then we write the machine ID to /etc/machine-id, and don't bother with /run/machine-id? hence what does this actually change here?

Reply:
I don't know about the initrd case, however I hit this on softreboot when /etc/machined is an empty read-only file, so we mount over it, but we lose this mount on softreboot and you get a different machine-id every time